src/main/java/org/broadinstitute/barclay/help/WDLWorkUnitHandler.java

```diff
@@ -183,71 +115,129 @@ protected void processPositionalArguments(
+                    // too, since they can't be required...
+                    optionalCompanions.add(companionMap);
+                } else {
+                    requiredCompanions.add(companionMap);
+                }
+                if (workflowInput.localizationOptional() == true) {
+                    localizationOptional.add(companionMap.get(TemplateProperties.ARGUMENT_NAME));
+                }
+            }
+
+            for (final String companion : workflowInput.optionalCompanions()) {
+                final Map<String, String> companionMap = createCompanionMapEntry(wdlName, companion);
+                optionalCompanions.add(companionMap);
+                if (workflowInput.localizationOptional() == true) {
+                    localizationOptional.add(companionMap.get(TemplateProperties.ARGUMENT_NAME));
+                }
+            }
         }
-        for (final String companion : workflowResource.companionResources()) {
-            final String companionArgOption = LONG_OPTION_PREFIX + companion;
-            companionFiles.merge(
-                    wdlName,
-                    Collections.singletonList(companionArgOption),
-                    (oldList, newList) -> {
-                        final List<String> mergedList = new ArrayList<>();
-                        mergedList.addAll(oldList);
-                        mergedList.add(companionArgOption);
-                        return mergedList;
-                    });
+
+        if (workflowOutput != null) {
+            for (final String companion : workflowOutput.requiredCompanions()) {
+                final Map<String, String> companionMap = createCompanionMapEntry(wdlName, companion);
+                runtimeOutputs.put(companionMap.get(TemplateProperties.ARGUMENT_NAME), wdlType);
+                if (resourceIsOptional) {
+                    optionalCompanions.add(companionMap);
+                } else {
+                    // required companions are only required if the source is required; if the source is
+                    // optional, then the companions have to be optional too, since they can't be required...
+                    requiredCompanions.add(companionMap);
+                    requiredOutputs.put(companionMap.get(TemplateProperties.ARGUMENT_NAME), wdlType);
+                }
+            }
+            for (final String companion : workflowOutput.optionalCompanions()) {
+                final Map<String, String> companionMap = createCompanionMapEntry(wdlName, companion);
+                runtimeOutputs.put(companionMap.get(TemplateProperties.ARGUMENT_NAME), wdlType);
+                if (!resourceIsOptional) {
+                    requiredOutputs.put(companionMap.get(TemplateProperties.ARGUMENT_NAME), wdlType);
+                    requiredCompanions.add(companionMap);
+                }
+                optionalCompanions.add(companionMap);
```

Comment:
Do you need to add in the optional companions here even if `resourceIsOptional` is true?  I'm a little confused about why you might potentially add the same `companionMap` twice - once to optional and once to required.

Reply:
Oh yeah, good catch. It is a bit confusing because the optional and required companions are mutually exclusive, but the optional and required outputs are not (this is purely by convention, because it helps make the template code less complex). I'll add some comments and maybe update the names to reenforce this.

Reply:
I fixed the bug here, and also updated many of the names in this method to be a bit more explicit. I also updated some residual references to `WorkflowResource` left over in this file.